### DISCUSSION TITLE
Make whisper transcribe numbers in the actual spoken words

### DISCUSSION THREAD

**Discussion by Thresher12** [openai/whisper#1041] [Q&A] [answered]
Hi, is there a way to get whisper to transcribe numbers the way they are actually spoken rather than just converting them to numeric format? **Yeah I know I can postprocess it but this is a suboptimal solution** since for example for something like 2015 there are multiple ways you can say it like 'two thousand and fifteen' or 'two thousand fifteen' or 'twenty fifteen' and whisper just automatically converts them to the same number while I need the actual spoken words. Theres obviously some conversion process going on here so is there a way to turn it off?

**Comment by jongwook** [comment, marked as answer]
It's not an explicit conversion but the model predicting the most likely textual output end-to-end. You can try the following which blocks all numeric tokens and encourages the model to transcribe in them literally.

```python
from whisper.tokenizer import get_tokenizer

tokenizer = get_tokenizer(multilingual=False)  # use multilingual=True if using multilingual model
number_tokens = [
    i 
    for i in range(tokenizer.eot)
    if all(c in "0123456789" for c in tokenizer.decode([i]).removeprefix(" "))
]

...

model.transcribe("audio.mp3", suppress_tokens=[-1] + number_tokens, ...)
```

**Reply by Thresher12** [reply]
Doesn't seem to work. The transcript just gives numbers as usual.

```
def Transcriber(audiofile):
    import whisper
    from whisper.tokenizer import get_tokenizer

    #encourage model to transcribe words literally
    tokenizer = get_tokenizer(multilingual=False)  # use multilingual=True if using multilingual model
    number_tokens = [
        i
        for i in range(tokenizer.eot)
        if all(c in "0123456789" for c in tokenizer.decode([i]).removeprefix(" "))
    ]

    model = whisper.load_model("base")
    answer = model.transcribe(audiofile, suppress_tokens=[-1] + number_tokens)

    print(answer)

```

**Reply by jongwook** [reply]
Did you change it to `multilingual=True`?

**Reply by Thresher12** [reply]
okay I didn't know base was multilingual it seems to work so far thanks

**Reply by orianemartin** [reply]
I tried this, but I get the following error

> line 125, in <listcomp>
>     if all(c in "0123456789" for c in tokenizer.decode([i]).removeprefix(" "))
> AttributeError: 'str' object has no attribute 'removeprefix' 

Could anyone help me figure out why? thank you!

**Reply by Warp-MFT** [reply]
"Python String removeprefix() function removes the prefix and returns the rest of the string. If the prefix string is not found, then it returns the original string. **It is introduced in Python 3.9.0 version.**"

**Reply by jongwook** [reply]
@orianemartin You can use `tokenizer.decode([i]).strip()` instead!

**Reply by orianemartin** [reply]
@Warp-MFT I am on 3.12.0
@jongwook thank you very much, it worked!

**Reply by ulatekh** [reply]
This can also be done for [faster-whisper](https://github.com/SYSTRAN/faster-whisper) and [insanely-fast-whisper](https://github.com/Vaibhavs10/insanely-fast-whisper); they only differ in how the tokenizer is found.

For faster-whisper, with a multilingual model: \
`tokenizer = faster_whisper.tokenizer.Tokenizer(tokenizer=model.hf_tokenizer, task="transcribe", language="en", multilingual=True)` \
With a monolingual model: \
`tokenizer = faster_whisper.tokenizer.Tokenizer(tokenizer=model.hf_tokenizer, multilingual=False)`

For insanely-fast-whisper: \
`tokenizer = pipe.tokenizer`

**Reply by kdcyberdude** [reply]
`blank_token_id` at `220` with token `" "` is also suppressed with above logic!!
It also suppress this token `50256` - `""` that's an empty token with no char - not sure what is the use of this in the vocab?

**Reply by grzegorz700** [reply]
@jongwook
After evaluating it, the original answer after decoding gave me 2 unexpected cases (space and empty or not for printing string). Snipped of the case:
```python
print(f"#{tokenizer.decode([number_tokens[10]])}#")
print(f"#{tokenizer.decode([number_tokens[-1]])}#")
```
Output:
```
# #
##
```

 To exclude it, we need to modify the code to:
```python
number_tokens = [
    i
    for i in range(tokenizer.eot)
    if (all(c in "0123456789" for c in tokenizer.decode([i]).removeprefix(" "))
        and len(tokenizer.decode([i]).strip()) > 0)
]
```

**Comment by lixikun** [comment]
@jongwook  Hi, I tried this method, but I got this error:

![image](https://user-images.githubusercontent.com/6120860/227778273-10c6a594-0fad-4915-bbc0-2a5d05ff0d18.png)

And I check the Tokenizer ,and the eot api return int ,int array. Is it correct?
```
    @cached_property
    def eot(self) -> int:
        return self.encoding.eot_token
```

**Reply by lixikun** [reply]
ok, it is sucessful after I modified the code from 
if all(c in "0123456789" for c in tokenizer.decode(i).removeprefix(" ")) to 
if all(c in "0123456789" for c in tokenizer.decode([i]).removeprefix(" "))

**Reply by jongwook** [reply]
Thanks! This happened because of a recent change in the tokenizer class

**Comment by Warp-MFT** [comment]
In case anyone is looking for how to do this for Hugging Face Transformers:
`generation_config = GenerationConfig.from_pretrained(whisper_model_name)`
 `generation_config.suppress_tokens += number_tokens`
...
`predicted_ids = whisper_model.generate(input_features, generation_config=generation_config, forced_decoder_ids=forced_decoder_ids)`

**Reply by asr-lord** [reply]
Could you share entire code? I can't reproduce it. Thank you

```python
import torch
from transformers import AutoModelForSpeechSeq2Seq, AutoProcessor, pipeline
from datasets import load_dataset


device = "cuda:0" if torch.cuda.is_available() else "cpu"
torch_dtype = torch.float16 if torch.cuda.is_available() else torch.float32

model_id = "openai/whisper-large-v3"

model = AutoModelForSpeechSeq2Seq.from_pretrained(
    model_id, torch_dtype=torch_dtype, low_cpu_mem_usage=True, use_safetensors=True
)
model.to(device)

processor = AutoProcessor.from_pretrained(model_id)
tokenizer=processor.tokenizer

number_tokens = [
    i 
    for i in range(tokenizer.eot)
    if all(c in "0123456789" for c in tokenizer.decode([i]).removeprefix(" "))
]

pipe = pipeline(
    "automatic-speech-recognition",
    model=model,
    tokenizer=processor.tokenizer,
    feature_extractor=processor.feature_extractor,
    max_new_tokens=128,
    chunk_length_s=30,
    batch_size=16,
    return_timestamps=True,
    torch_dtype=torch_dtype,
    device=device,
)

result = pipe("test.mp3")
```
Error message:
```
      1 number_tokens = [
      2     i 
----> 3         for i in range(tokenizer.eot)
      4             if all(c in "0123456789" for c in tokenizer.decode([i]).removeprefix(" "))
      5             ]

AttributeError: 'WhisperTokenizer' object has no attribute 'eot'
```

**Reply by Warp-MFT** [reply]
It's your lucky day :) 
[open_ai_stt.txt](https://github.com/openai/whisper/files/15015690/open_ai_stt.txt)

**Reply by JabblyApp** [reply]
Hey, any doc for this for Javascript, I'm using this models with APIs and the numbers aren't transcribed, thanks !
